src/qibo/quantum_info/quantum_networks.py

```diff
@@ -0,0 +1,644 @@
+        """Returns bool indicating if Choi operator :math:`\\mathcal{E}` of the networn is positive-semidefinite.
+
+        Args:
+            precision_tol (float, optional): threshold value used to check if eigenvalues of
+                the Choi operator :math:`\\mathcal{E}` are such that
+                :math:`\\textup{eigenvalues}(\\mathcal{E}) >= - \\textup{precision_tol}`.
+                Note that this parameter can be set to negative values.
+                Defaults to :math:`0.0`.
+
+        Returns:
+            bool: Positive-semidefinite condition.
+        """
+        self._matrix = self._full()
+        self._pure = False
+
+        reshaped = np.reshape(self._matrix, (self.dims, self.dims))
+
+        if self.hermitian():
+            eigenvalues = np.linalg.eigvalsh(reshaped)
+        else:
+            if self._backend.__class__.__name__ in [
+                "CupyBackend",
+                "CuQuantumBackend",
+            ]:  # pragma: no cover
+                reshaped = np.array(reshaped.tolist(), dtype=reshaped.dtype)
+            eigenvalues = np.linalg.eigvals(reshaped)
+
+        return all(eigenvalue >= -precision_tol for eigenvalue in eigenvalues)
+
+    def channel(
+        self,
+        order: Optional[Union[int, str]] = None,
+        precision_tol_causal: float = 1e-8,
+        precision_tol_psd: float = 1e-8,
+    ):
+        """Returns bool indicating if Choi operator :math:`\\mathcal{E}` is a channel.
+
+        Args:
+            order (int or str, optional): order of the norm used to calculate causality.
+                Defaults to ``None``.
+            precision_tol_causal (float, optional): threshold :math:`\\epsilon` that defines if
+                Choi operator of the network is :math:`\\epsilon`-close to causality in the norm
+                given by ``order``. Defaults to :math:`10^{-8}`.
+            precision_tol_psd (float, optional): threshold value used to check if eigenvalues of
+                the Choi operator :math:`\\mathcal{E}` are such that
+                :math:`\\textup{eigenvalues}(\\mathcal{E}) >= \\textup{precision_tol_psd}`.
+                Note that this parameter can be set to negative values.
+                Defaults to :math:`0.0`.
+
+        Returns:
+            bool: Channel condition.
+        """
+        return self.causal(order, precision_tol_causal) and self.positive_semidefinite(
+            precision_tol_psd
+        )
+
+    def apply(self, state):
+        """Apply the Choi operator :math:`\\mathcal{E}` to ``state`` :math:`\\varrho`.
+
+        It is assumed that ``state`` :math:`\\varrho` is a density matrix.
+
+        Args:
+            state (ndarray): density matrix of a ``state``.
+
+        Returns:
+            ndarray: Resulting state :math:`\\mathcal{E}(\\varrho)`.
+        """
+        matrix = np.copy(self._matrix)
+
+        if self.pure():
+            return np.einsum("jk,lm,jl -> km", matrix, np.conj(matrix), state)
```

Comment:
@Canoming let me ask you something: this method, when ``self.pure()`` is ``True``, is applying $U^{T} \\, \rho \\, U^{*}$ instead of the traditional $U \\, \rho \\, U^{\\dagger}$. Is that supposed to be the case or is it a bug?